core/src/main/java/io/grpc/internal/TimeProvider.java

```diff
@@ -29,7 +29,10 @@ public interface TimeProvider {
   TimeProvider SYSTEM_TIME_PROVIDER = new TimeProvider() {
     @Override
     public long currentTimeNanos() {
-      return TimeUnit.MILLISECONDS.toNanos(System.currentTimeMillis());
+      Instant instant = Instant.now();
+      int nanos = instant.getNano();
+      long epochSeconds = instant.getEpochSecond();
+      return Long.parseLong("" + epochSeconds + nanos);
```

Comment:
This is not correct. We want the value in nanos. Epoch second of 1 (for example) and Instant.getNano of 1 as per your logic is 11 epoch nanos, whereas it should be 1000000001.

Reply:
We also need to write a unit test for this.

Reply:
Fixed along with adding junits